Fix flaky encryption tests

p8 opened the pull request:
`4f365720d1b87b7a8fe7d8d77c157b63d853e334` introduced lazy loading
encryption schemes. Schemes for attributes are loaded when the attribute
type is used.

This change introduced some flakiness to the tests:

```bash
ARCONN=mysql2 bin/test test/cases/encryption/encryption_schemes_test.rb test/cases/encryption/encrypted_fixtures_test.rb -n "/^(?:ActiveRecord::Encryption::EncryptionSchemesTest#(?:test_returns_ciphertext_all_the_previous_schemes_fail_to_decrypt_and_support_for_unencrypted_data_is_on)|ActiveRecord::Encryption::EncryptableFixtureTest#(?:test_preserved_columns_due_to_ignore_case:_true_gets_encrypted_automatically))$/" --seed 39616

...

.E

 Error:
    ActiveRecord::Encryption::EncryptableFixtureTest#test_preserved_columns_due_to_ignore_case:_true_gets_encrypted_automatically:
    ActiveRecord::Encryption::Errors::Decryption: Couldn't find a match for ruby for rails ({"1"=>"2"})
        /Users/petrik/Projects/All/_forks/rails/activerecord/test/cases/encryption/encryption_schemes_test.rb:144:in `decrypt'
```


By calling the attribute type in the test models the schemes are eager
loaded and the flakiness is hopefully fixed. (The original commit added
a similar call for the type).

cc: @jorgemanrubia 

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil commented:
Does this address #48605?

p8 commented:
> Does this address #48605?

It does for me 😄

p8 commented:
It's a bit weird though because the schemes are supposed to be cleared in the default setup:
https://github.com/rails/rails/blob/c388a4dcf580281aef4e341f87c52956b0b0057a/activerecord/test/cases/encryption/helper.rb#L171

p8 commented:
hmm, it still fails for some seeds it seems :(

yahonda commented:
Thanks for working on this. Here are the minimum cases to reproduce failures.

https://buildkite.com/rails/rails/builds/97759#01891d7b-d637-4c0b-8edf-44ad788dae6e/1164-1173

```ruby
$ ARCONN=mysql2 bin/test test/cases/encryption/configurable_test.rb --seed 48114
Using mysql2
Run options: --seed 48114

# Running:

F

Failure:
ActiveRecord::Encryption::ConfigurableTest#test_.configure_configures_initial_config_properties [/home/yahonda/src/github.com/rails/rails/activerecord/test/cases/encryption/configurable_test.rb:26]:
--- expected
+++ actual
@@ -1 +1 @@
-#<ActiveRecord::Encryption::DerivedSecretKeyProvider:0xXXXXXX @keys=[#<ActiveRecord::Encryption::Key:0xXXXXXX @secret="\x84\xB5^\x14y\xE0\xD7\x88\x062\x02\x10o\xCC\x9C\"\xC7[\xAF\xBAI6\xBCj\xB0R\xCCQ~\xD6\xC7\xED", @public_tags=#<ActiveRecord::Encryption::Properties:0xXXXXXX @data={}>>]>
+#<ActiveRecord::Encryption::DerivedSecretKeyProvider:0xXXXXXX @keys=[#<ActiveRecord::Encryption::Key:0xXXXXXX @secret="T\xE0\xA3?\xCENQ7u;\xFE\xD0\xFD\xA8-\xEB\xFE^=\xE8G\xC8v\xF3\xD0\xEC[X6\x05\xD4e", @public_tags=#<ActiveRecord::Encryption::Properties:0xXXXXXX @data={}>>]>



bin/test test/cases/encryption/configurable_test.rb:12

.....

Finished in 0.120896s, 49.6296 runs/s, 99.2592 assertions/s.
6 runs, 12 assertions, 1 failures, 0 errors, 0 skips
$
```

https://buildkite.com/rails/rails/builds/97759#01891d7b-d6e1-4f20-9545-2118fe9e086f/1077-1086

```ruby
$ ARCONN=mysql2 bin/test test/cases/autosave_association_test.rb test/cases/encryption/uniqueness_validations_test.rb -n "/^(?:TestAutosaveAssociationValidationsOnAHasManyAssociation#(?:test_should_automatically_validate_associations)|ActiveRecord::Encryption::UniquenessValidationsTest#(?:test_uniqueness_validations_work_when_using_old_encryption_schemes))$/"  --seed 40224
Using mysql2
Run options: -n "/^(?:TestAutosaveAssociationValidationsOnAHasManyAssociation#(?:test_should_automatically_validate_associations)|ActiveRecord::Encryption::UniquenessValidationsTest#(?:test_uniqueness_validations_work_when_using_old_encryption_schemes))$/" --seed 40224

# Running:

F

Failure:
ActiveRecord::Encryption::UniquenessValidationsTest#test_uniqueness_validations_work_when_using_old_encryption_schemes [/home/yahonda/src/github.com/rails/rails/activerecord/test/cases/encryption/uniqueness_validations_test.rb:37]:
ActiveRecord::RecordInvalid expected but nothing was raised.


bin/test test/cases/encryption/uniqueness_validations_test.rb:25

.

Finished in 0.182109s, 10.9824 runs/s, 16.4737 assertions/s.
2 runs, 3 assertions, 1 failures, 0 errors, 0 skips
$
```

p8 commented:
Thanks @yahonda those really help.
I've bisected this to commit: 4f365720d1b87b7a8fe7d8d77c157b63d853e334

p8 committed:
Fix flaky encryption tests

`4f365720d1b87b7a8fe7d8d77c157b63d853e334` introduced lazy loading
encryption schemes. Schemes for attributes are loaded when the attribute
type is used.

This change introduced some flakiness to the tests:

```bash
ARCONN=mysql2 bin/test test/cases/encryption/encryption_schemes_test.rb test/cases/encryption/encrypted_fixtures_test.rb -n "/^(?:ActiveRecord::Encryption::EncryptionSchemesTest#(?:test_returns_ciphertext_all_the_previous_schemes_fail_to_decrypt_and_support_for_unencrypted_data_is_on)|ActiveRecord::Encryption::EncryptableFixtureTest#(?:test_preserved_columns_due_to_ignore_case:_true_gets_encrypted_automatically))$/" --seed 39616

...

.E

Error:
ActiveRecord::Encryption::EncryptableFixtureTest#test_preserved_columns_due_to_ignore_case:_true_gets_encrypted_automatically:
ActiveRecord::Encryption::Errors::Decryption: Couldn't find a match for ruby for rails ({"1"=>"2"})
    /Users/petrik/Projects/All/_forks/rails/activerecord/test/cases/encryption/encryption_schemes_test.rb:144:in `decrypt'
```

By calling the attribute type in the test models the schemes are eager
loaded and the flakiness is hopefully fixed. (The original commit added
a similar call for the type).

Co-authored-by: Yasuo Honda <yasuo.honda@gmail.com>

yahonda commented on activerecord/test/cases/encryption/encryption_schemes_test.rb:
```diff
@@ -179,6 +179,6 @@ def declare_class_with_global_previous_encryption_schemes(*previous_schemes)
 
         encrypts :name
-      end
+      end.tap {|c| c.type_for_attribute(:name) }
```
```suggestion
      end.tap { |c| c.type_for_attribute(:name) }
```

jorgemanrubia reviewed (commented):
@p8 thank you so much for tracking this one down and fixing 🙏. I totally missed this. Let me see if I can prepare a more seamless patch here before merging this one please.

jorgemanrubia commented:
I just created #48659, which aims to do the same you did here automatically @p8. I added you as coauthor there, you did all the work here 🙏.

p8 commented:
Closing in favor of: https://github.com/rails/rails/pull/48659
